feat: support more fields to sortBy in /api/v1/experiments/trials [DET-4219, DET-4920]

stoksc opened the pull request:
## Description
This change adds support to sort by validation end time, validation metric, duration and total batches processed to `/api/v1/experiments/$1/trials`. It also fixes a bug; it seems none of the sort by's worked previously (this can be verified on latest master). Also, it adds the SQL function `public.page_info` that replicates the behavior of `api.Pagination()` in Go code and fixes a regression in `/api/v1/experiments` where using offset and limit as provided didn't work as expected (e.g. GET `/api/v1/experiments` with no query params returned `[]` on a full cluster).
<!---
Lead with the intended commit body in this description field. For breaking changes,
please include "BREAKING CHANGE:" at the beginning of your commit body.
At minimum, this section should include a bracketed reference to the Jira ticket,
e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.
-->


## Test Plan
- [x] hit all the endpoints a lot with swagger
- [ ] maybe add integrations if i get a second
<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.
-->



<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

stoksc commented on master/internal/db/postgres.go:
```diff
@@ -1571,6 +1571,38 @@ WHERE id = :id`, setClause(toUpdate)), validation)
 			strings.Join(toUpdate, ", "), trialID, stepID)
 	}
+
+	if validation.State == model.CompletedState {
```
This is the only part I'm not really in love with, those sort by best validation is a very common action and scanning the trials table and the validations table for each trial to order them all is probably ridiculously expensive, so somewhere like this is where we're at.

shiyuann commented on master/static/srv/proto_get_trial_ids_for_experiment.sql:
```diff
@@ -0,0 +1,66 @@
+                AND s.state = 'COMPLETED'
+            ORDER BY s.id DESC
+            LIMIT 1
+        ) AS total_batches_processed,
```
:) wondering if we should also make this a separate column.

stoksc replied:
heh. yeah, sgtm!

shiyuann commented on master/static/srv/proto_get_trial_ids_for_experiment.sql:
```diff
@@ -0,0 +1,66 @@
+    (SELECT coalesce(json_agg(paginated_experiment_trials), '[]'::json) FROM (
+        SELECT id FROM filtered_experiment_trials
+        ORDER BY %s
+        OFFSET (SELECT p.page_info->>'start_index' FROM page_info p)::bigint
```
nonblocking: Why not just use `$3` or `$4` for OFFSET and LIMIT?

stoksc replied:
If you just use `$4` for limit then if it's zero we dont default it. latest master `GET /api/v1/experiments` with no params is returning `[]` which is a bug. Offset has some similar problems. Also this is partially to keep the same pagination behavior as before.

stoksc replied:
Really, you could default limit outside and just use them - the other reason I was avoid that was it was nice this way to return the full `apiv1.Pagination` json so we can just `QueryProtof` into the resp object.

shiyuann commented on master/static/srv/proto_get_trial_ids_for_experiment.sql:
```diff
@@ -0,0 +1,66 @@
+                 ORDER BY searcher_info.sign * (v.metrics->'validation_metrics'->>searcher_info.metric_name)::float8 ASC
+                 LIMIT 1)
+            END
+        ) as best_signed_search_metric,
```
Is it used somewhere?

shiyuann commented on master/static/srv/proto_get_trial_ids_for_experiment.sql:
```diff
@@ -0,0 +1,66 @@
+           WHERE v.end_time IS NOT NULL
+           ORDER BY v.id DESC
+           LIMIT 1
+        ) as validation_end_time
```
same question?

stoksc replied:
this and above are used in the order by

shiyuann commented on master/internal/api_experiment.go:
```diff
@@ -121,60 +120,19 @@ func (a *apiServer) GetExperiments(
 		labelFilterExpr,
 		req.Description,
 		req.Offset,
 		req.Limit,
```
BTW, do you mind setting the limit to be 100 here? That fix for [DET-4917](https://determinedai.atlassian.net/browse/DET-4917) also needs to get into the next release.

stoksc replied:
It's fixed already, but in the `public.page_info` function. It defaults to all like it used to.

shiyuann reviewed (commented):
I saw you have put a file `proto/buf.image.bin` in the PR. Isn't it a generated binary file?

stoksc commented on master/static/migrations/20210129225517_add-validation-columns.up.sql:
```diff
@@ -0,0 +1,26 @@
+ALTER TABLE public.trials
+    ADD COLUMN best_validation_id int NULL;
```
I'd rather leave it as the ID; there are other places that use this and need the whole (see `proto_get_trials_plus` which gets the trial's best validation and returns it).

stoksc commented:
> I saw you have put a file proto/buf.image.bin in the PR. Isn't it a generated binary file?

Yeah, initially I had a breaking change as part of this, but I reverted it (I removed sort by state originally). I'll revert that.

shiyuann commented on master/internal/db/postgres.go:
```diff
@@ -72,7 +72,7 @@ func (db *PgDB) Migrate(migrationURL string) error {
 
 	if err = m.Up(); err != migrate.ErrNoChange {
-		return errors.Wrap(err, "error applying migrations")
+		return errors.Wrap(err, "d")
```
```suggestion
		return errors.Wrap(err, "error applying migrations")
```

stoksc replied:
oops.. good catch.

shiyuann reviewed (approved):
LGTM!
